proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+None of these sufficiently express cryptography through the type-system and with type-classes. `crypton/ite` exposes a few highly- opinionated / specific classes for hashing, block and stream ciphers, but otherwise is focused on concrete implementations and functions. Both `crypton` and `cryptonite` rely on bindings to C implementations of unknown veracity; though the code is itself not suspect, the lack of provenance makes it ill-suited for recommendation in trusted environments, and maintenance of the C implementation poses burden of high technical skill. Libsodium bindings are more trustworthy, but the limited options severely hamper interop with other systems unless they too are using libsodium, making them ill-suited for recommendation for some use cases. `Z-Botan` has been apparently dead for several years, and even if revived has a has a sizeable dependency in the form of the `Z-Haskell` ecosystem.
+
+Furthermore, none of these libraries contain implementations of post-quantum cryptography schemes; although existing quantum computers are of insufficient capability to break commonly used algorithms such as Ed25519, this will not remain true forever, and it would be best to have quantum-resistant implementations ready for adoption long before such attacked become practical.
+
+Each of these efforts has one or more of these issues that preclude them from being considered a complete success:
+
+- Insufficient abstraction
+- Maintenance of handwritten C
+- Bindings to an insufficiently general library
+- Burdensome dependencies
+- Dead project
+- Lack of post-quantum algorithms
+
+However, there is also a lot to be learned from the ways these prior efforts *were* successful. This proposal will assess these existing efforts in an attempt to consolidate their learnings and functionality.
+
+## An example of a successful ecosystem
+
+The health of the Haskell server ecosystem may be used as an example of a successful community-driven niche, and we can aspire to bring that same quality and health to the Haskell cryptography ecosystem. There are many pertinent parallels in terms of managing performance, complexity, and safety that we can learn from:
+
+- Low-level packages like `wai` supporting multiple backends
+- Backends range from simple like `scotty` to complex like `servant`
+- Frontends like `blaze-html` and `lucid`
+- Libraries can take advantage of type system to provide good abstractions
+- Flexible enough abstractions to explore the problem space with multiple solutions
+- Appropriate licensing
```

Comment:
Licensing keeps coming up - it would also be good to be more explicit about your background assumptions, which are presumably that business users are likely to be unwilling to depend on copyleft code, and we want to provide value to them too.

Reply:
Agreed.